From 8ec9a94bb8f56457ec6dc0ea7e51edb20c74b493 Mon Sep 17 00:00:00 2001
From: Vinith Kumar R <quic_vinithku@quicinc.com>
Date: Thu, 30 Nov 2023 10:30:46 +0530
Subject: [PATCH] wifi: ath12k: Avoid wrong tx completion notification to
 mac80211

Currently few exception packets we receive tx completion from firmware
and the tx status updated to mac80211. Currently tx status values
HAL_WBM_REL_HTT_TX_COMP_STATUS_DROP and HAL_WBM_REL_HTT_TX_COMP_STATUS_TTL
are considered as tx failure and reported as tx failure to mac80211
which triggers mpath broken issue due to increase in mpath fail
average. But these two reasons are not tx failure its dropped in
the hardware due to queue is full and exceeded the time to live
threshold, and these packets were not transmitted in the air.
So do not report these tx status to mac80211 and just free
the skb in driver to avoid invalid mpath broken issue.

mesh mpl broken link print:
2023-11-07 17:36:40 :: [406374.479678] wlan1-15:  MESH MPL link to
3a:03:7f:85:66:17 is broken and 1 path deactivated signal -12 dbm
signal_avg -12 dbm medium_busy : 0

Signed-off-by: Vinith Kumar R <quic_vinithku@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index 00be3da..7523e28 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -964,8 +964,6 @@ ath12k_dp_tx_process_htt_tx_complete(struct ath12k_base *ab,
 	ab->soc_stats.fw_tx_status[htt_status]++;
 
 	switch (htt_status) {
-	case HAL_WBM_REL_HTT_TX_COMP_STATUS_DROP:
-	case HAL_WBM_REL_HTT_TX_COMP_STATUS_TTL:
 	case HAL_WBM_REL_HTT_TX_COMP_STATUS_OK:
 		ts.acked = (htt_status == HAL_WBM_REL_HTT_TX_COMP_STATUS_OK);
 		ts.ack_rssi = le32_get_bits(status_desc->info2,
@@ -973,6 +971,8 @@ ath12k_dp_tx_process_htt_tx_complete(struct ath12k_base *ab,
 		ath12k_dp_tx_htt_tx_complete_buf(ab, msdu, tx_ring, &ts, skb_ext_desc, mac_id, desc, timestamp);
 
 		break;
+	case HAL_WBM_REL_HTT_TX_COMP_STATUS_DROP:
+	case HAL_WBM_REL_HTT_TX_COMP_STATUS_TTL:
 	case HAL_WBM_REL_HTT_TX_COMP_STATUS_REINJ:
 	case HAL_WBM_REL_HTT_TX_COMP_STATUS_INSPECT:
 	case HAL_WBM_REL_HTT_TX_COMP_STATUS_VDEVID_MISMATCH:
-- 
2.17.1

